adcc/IsrMatrix.py

```diff
@@ -0,0 +1,158 @@
+        self.method = method
+        self.ground_state = hf_or_mp
+        self.reference_state = hf_or_mp.reference_state
+        self.mospaces = hf_or_mp.reference_state.mospaces
+        self.is_core_valence_separated = method.is_core_valence_separated
+        self.ndim = 2
+        self.extra_terms = []
+
+        # Determine orders of PT in the blocks
+        if block_orders is None:
+            block_orders = self.default_block_orders[method.base_method.name]
+        else:
+            tmp_orders = self.default_block_orders[method.base_method.name].copy()
+            tmp_orders.update(block_orders)
+            block_orders = tmp_orders
+
+        # Sanity checks on block_orders
+        for block in block_orders.keys():
+            if block not in ("ph_ph", "ph_pphh", "pphh_ph", "pphh_pphh"):
+                raise ValueError(f"Invalid block order key: {block}")
+        if block_orders["ph_pphh"] != block_orders["pphh_ph"]:
+            raise ValueError("ph_pphh and pphh_ph should always have "
+                             "the same order")
+        if block_orders["ph_pphh"] is not None \
+           and block_orders["pphh_pphh"] is None:
+            raise ValueError("pphh_pphh cannot be None if ph_pphh isn't.")
+        self.block_orders = block_orders
+
+        # Build the blocks
+        with self.timer.record("build"):
+            variant = None
+            if self.is_core_valence_separated:
+                variant = "cvs"
+            blocks = [{
+                block: ppbmatrix.block(self.ground_state, dip, block.split("_"),
+                                       order=order, variant=variant)
+                for block, order in self.block_orders.items() if order is not None
+            } for dip in self.dips]
+            # TODO Rename to self.block in 0.16.0
+            self.blocks_ph = [{
+                b: bl[b].apply for b in bl
+            } for bl in blocks]
+
+    @timed_member_call()
+    def matvec(self, v):
+        """
+        Compute the matrix-vector product of the ISR one-particle
+        operator and return the result.
+        """
+        ret = [
+            sum(block(v) for block in bl_ph.values())
+            for bl_ph in self.blocks_ph
+        ]
+        if len(ret) == 1:
+            return ret[0]
+        else:
+            return ret
+
+    def rmatvec(self, v):
+        # Hermitian operators
+        if all(dip.is_symmetric for dip in self.dips):
+            return self.matvec(v)
+        else:
+            diffv = [dip.ov + dip.vo.transpose((1, 0)) for dip in self.dips]
+            # anti-Hermitian operators
+            if all(dv.dot(dv) < 1e-12 for dv in diffv):
+                return [
+                    AmplitudeVector(ph=-1.0 * mv.ph, pphh=-1.0 * mv.pphh)
+                    for mv in self.matvec(v)
+                ]
+            # operators without any symmetry
+            else:
+                return NotImplemented
+
+    def __matmul__(self, other):
```

Comment:
Can you please add a docstring about what happens when you apply the matrix with multiple operators to multiple vectors? Do I get a 2D list with N_operators x N_vectors back?

Reply:
Done! No, you get a 2D list with N_vectors x N_operators back.